internal/cluster/watch.go

```diff
@@ -0,0 +1,204 @@
+package cluster
+
+import (
+	"fmt"
+	"github.com/jenkins-x-apps/jx-app-jacoco/internal/report"
+	"github.com/jenkins-x-apps/jx-app-jacoco/internal/util"
+	"github.com/jenkins-x/jx/pkg/kube"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/tools/cache"
+	"time"
+
+	jenkinsv1 "github.com/jenkins-x/jx/pkg/apis/jenkins.io/v1"
+	jenkinsclientv1 "github.com/jenkins-x/jx/pkg/client/clientset/versioned/typed/jenkins.io/v1"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	logger = log.WithFields(log.Fields{"app": "jacoco"})
+)
+
+// WatchPipelineActivity watches the jx namespace for changes to PipelineActivities.
+func WatchPipelineActivity(namespace string, jxClient *jenkinsclientv1.JenkinsV1Client) {
+	listWatch := cache.NewListWatchFromClient(jxClient.RESTClient(), "pipelineactivities", namespace, fields.Everything())
+	kube.SortListWatchByName(listWatch)
+	_, actController := cache.NewInformer(
+		listWatch,
+		&jenkinsv1.PipelineActivity{},
+		time.Minute*10,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				onPipelineActivity(obj, jxClient)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				onPipelineActivity(newObj, jxClient)
+			},
+			DeleteFunc: func(obj interface{}) {},
+		},
+	)
+	stop := make(chan struct{})
+	go actController.Run(stop)
+}
+
+func onPipelineActivity(obj interface{}, jxClient *jenkinsclientv1.JenkinsV1Client) {
+	pipelineActivity, ok := obj.(*jenkinsv1.PipelineActivity)
+	if !ok {
+		logger.Warnf("unexpected type %T", obj)
+	} else {
+		err := handlePipelineActivity(pipelineActivity, jxClient)
+		if err != nil {
+			logger.Error(err)
+		}
+	}
+}
+
+func handlePipelineActivity(pipelineActivity *jenkinsv1.PipelineActivity, jxClient *jenkinsclientv1.JenkinsV1Client) (err error) {
+	for _, attachment := range pipelineActivity.Spec.Attachments {
+		if attachment.Name != "jacoco" {
+			continue
+		}
+
+		// TODO Handle having multiple attachments properly
+		for _, url := range attachment.URLs {
+			urlWithTimestamp := fmt.Sprintf("%s?version=%d", url, time.Now().UnixNano()/int64(time.Millisecond))
+
+			if containsFactForURL(pipelineActivity, url) {
```

Comment:
So this is a bit new. We also check whether there is already a fact for this URL and if so skip processing. Otherwise we keep updating over and over with the same data. Really the XML report should not change. If it does, there is something wrong (eg manual change).